### ISSUE TITLE
Running a container that hogs disk I/O write throughput hurts daemon performance

### ISSUE THREAD

**Issue by kangp3** [moby/moby#39737]
**Description**

One of the containers that we run is a pretty huge disk I/O hog. It's essentially a daemon that is responsible for downloading a set of large (~100MB) files in parallel and caching them to disk. Because of this, it quickly saturates the ~50MB/s maximum write throughput of the instances we're running them on (VMs in GCP). While the container is running, the daemon performance degrades considerably when we attempt to create or start containers, with each request taking upwards of tens of seconds, even over a minute in the worst case.

In order to mitigate this problem, we attempted to apply the `--device-write-bps` flag as well as the `BlkioDeviceWriteBps` field of our host configs in Go. This may be a separate issue, but for our specific use-case we found that setting the flag was not having any effect. Write throughput didn't appear to be throttled, and we were still experiencing issues with degraded daemon performance.

In summary, this issue kind of encapsulates two issues:
1. Daemon performance for container creation/start is degraded when one of the containers is an disk write throughput hog.
    - This issue is definitely expected. If there's an easy solution we're missing would love to hear it though 😄
2. The write throughput throttles don't seem to apply to some containers. I've tested it with some toy examples that work and some that don't. See repro steps below for details.
    - This issue I could definitely use some help with. This seems like it would be our silver bullet, but because it's not working we've been forced to investigate other alternatives to quarantine the I/O hogging of our container.

**Steps to reproduce the issue:**
1. Run the following container:
    ```
    $ docker run --rm --device-write-iops /dev/sda:50 --device-write-bps /dev/sda:10mb --name io_rate_limit ubuntu:xenial bash -c 'apt-get update
    apt-get install -y openssl
    dd if=/dev/urandom bs=128 count=1 | base64 > /passfile
    echo "starting write"
    dd if=/dev/zero bs=1M count=10000 | openssl enc -aes-256-ctr -pass file:passfile -nosalt > testfile'
    ```
    - This puts up an Ubuntu container that is being throttled at 50 write IOPS/s and 10MB/s that does the following:
        1. Install `openssl` (this is later used to produce a file with pseudo-random contents by encrypting `/dev/zero`, which is faster than reading `/dev/urandom`)
        2. Create a `/passfile` that is used for the encryption
        3. Start encrypting `/dev/zero` to produce pseudo-random content and write it to a file
2. Monitor host I/O usage with `iotop`
3. Attempt to interact with the daemon after the last part of the command has started running

**Describe the results you received:**
- During the running of the container, the Docker daemon becomes sluggish as described. Docker container delete/create/start requests all take some time to go through
- The `dd` command reports that it wrote at well over the throttled limit of 10MB/s, although this might be some side-effect of the way that `dd` is calculating write speed
    ```
    starting write               
    5000+0 records in            
    5000+0 records out           
    5242880000 bytes (5.2 GB, 4.9 GiB) copied, 54.6254 s, 96.0 MB/s
    ```
- The `iotop` monitoring reports I/O usage of well over the throttled limits, peaking at ~150MB/s and steady-stating at the estimated sustained disk throughput limit of ~50MB/s. I/O percent usage seems to be dominated by three commands, namely `kworker`, `jbd2`, and the expected `openssl`. This seems related to https://github.com/moby/moby/issues/35274, but this time I'm not using mysql to my knowledge.
    ```
    Total DISK READ :       0.00 B/s | Total DISK WRITE :      51.18 M/s
    Actual DISK READ:       0.00 B/s | Actual DISK WRITE:      59.22 M/s
      TID  PRIO  USER     DISK READ  DISK WRITE  SWAPIN     IO>    COMMAND
    28823 be/4 root        0.00 B/s    0.00 B/s  0.00 % 99.99 % [kworker/u4:0]
      327 be/3 root        0.00 B/s    0.00 B/s  0.00 % 96.72 % [jbd2/sda1-8]
    31687 be/4 root        0.00 B/s   51.18 M/s  0.00 % 83.40 % openssl enc -aes-256-ctr -pass file:passfile -nosalt
    ```
- Requests to the daemon to create/delete containers are quite sluggish
    ```
    $ time docker run ubuntu echo hi
    hi

    real    2m47.051s
    user    0m0.071s
    sys     0m0.054s
    ```
    ```
    $ time docker rm -f io_rate_limit
    io_rate_limit

    real    0m13.636s
    user    0m0.031s
    sys     0m0.034s
    ```
    - Note: Immediately after the `rm -f` command, the `openssl` process goes down but `kworker` and `jbd2` continue to eat up I/O for a few seconds before calming down and allowing the remove operation to complete.

**Describe the results you expected:**
- With the rate limiting, I would expect the container to not take up the I/O it seems to take up. I would expect both the `dd` and `iotop` outputs to reflect the disk write throttle of 10MB/s. Ideally, with the disk less busy, this would also mean that the daemon wouldn't be as choked up and Docker would continue running smoothly.

**Additional information you deem important (e.g. issue happens only occasionally):**
- With my limited knowledge of cgroups I poked around the container's cgroup and found that the container PIDs were correctly being listed. This makes me believe that the `kworker` and `jbd2` processes have more to do with it, but my understanding of those two processes is limited. From the issue I mentioned before it seems like this could be mitigated by disabling journaling on the disk somehow? Unfortunately I've been running this on the root volume, and disabling journaling there seems nontrivial.
    ```
    $ cat /sys/fs/cgroup/blkio/docker/8d219f950f3a137e839a28d13df221abf0abc46c20fac1f366eca04576c8e4e7/cgroup.procs
    1396
    1720
    1721
    $ ps aux | grep -e 1396 -e 1720 -e 1721
    root      1396  0.1  0.0  18024  3020 ?        Ss   21:52   0:00 bash -c apt-get update apt-get install -y openssl dd if=/dev/urandom     bs=128 count=1 | base64 > /passfile echo "starting write" dd if=/dev/zero bs=1M count=10000 | openssl enc -aes-256-ctr -pass     file:passfile -nosalt > testfile
    root      1720 21.5  0.0   5452  2288 ?        S    21:52   0:05 dd if=/dev/zero bs=1M count=10000
    root      1721 38.0  0.0  13940  3540 ?        D    21:52   0:10 openssl enc -aes-256-ctr -pass file:passfile -nosalt
    ubuntu    1744  0.0  0.0  12944   968 pts/3    S+   21:53   0:00 grep --color=auto -e 1396 -e 1720 -e 1721    
    ```

**Output of `docker version`:**

```
$ docker version
Client:
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        6247962
 Built:             Sun Feb 10 04:13:50 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 03:42:13 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
$ docker info
Containers: 4
 Running: 1
 Paused: 0
 Stopped: 3
Images: 18
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: nvidia runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-1027-gcp
Operating System: Ubuntu 16.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.298GiB
Name: local-peter-cpu-176267314
ID: 6H6G:DIQG:PZIL:6MVS:KHQG:3VLT:BTOI:AH6B:K2PL:7QVL:43DN:E722
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 REDACTED
Registry Mirrors:
 REDACTED
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
GCP VM instance
    - n1-standard-2 (2 vCPUs, 7.5 GB memory)
    - 110 GB SSD Persistent Disk
    - Disk estimated sustained random IOPS limit (taken from GCP console): 3300/3300 R/W
    - Disk estimated sustained throughput limit (MB/s) (take from GCP console): 52.8/52.8 R/W
